kubernetes/nodecollection.go

```diff
@@ -28,35 +35,136 @@ func NewClientsetNodeSource(clientSet kubernetes.Interface) ClientsetNodeSource
-func (cns ClientsetNodeSource) GetNodes() ([]string, error) {
-	nodeObjs, err := cns.clientSet.CoreV1().Nodes().List(metav1.ListOptions{})
-	nodeNames := make([]string, 0, len(nodeObjs.Items))
-	for _, node := range nodeObjs.Items {
-		nodeNames = append(nodeNames, node.Name)
+func (cns ClientsetNodeSource) GetNodes() (*v1.NodeList, error) {
+	return cns.clientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+}
+
+// NodeAddress returns the internal IP address and kubelet port of a given node
+func (cns ClientsetNodeSource) NodeAddress(node *v1.Node) (string, int32, error) {
+	// adapted from k8s.io/kubernetes/pkg/util/node
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == v1.NodeInternalIP {
+			return addr.Address, node.Status.DaemonEndpoints.KubeletEndpoint.Port, nil
+		}
 	}
-	return nodeNames, err
+	return "", 0, fmt.Errorf("Could not find internal IP address for node %s ", node.Name)
 }
 
 func downloadNodeData(prefix string,
 	config KubeAgentConfig,
 	workDir *os.File,
-	rawClient raw.Client,
-	nodeSource NodeSource) error {
+	nodeSource NodeSource) (map[string]error, error) {
+
+	failedNodeList := make(map[string]error)
 
 	nodes, err := nodeSource.GetNodes()
 
 	if err != nil {
-		return fmt.Errorf("cloudability metric agent is unable to get a list of nodes: %v", err)
+		return nil, fmt.Errorf("cloudability metric agent is unable to get a list of nodes: %v", err)
 	}
 
-	for _, n := range nodes {
+	for _, n := range nodes.Items {
+		// retrieve node summary directly from node
+		if config.nodeRetrievalMethod == "direct" {
+
+			ip, port, err := nodeSource.NodeAddress(&n)
+			if err != nil {
+				return nil, fmt.Errorf("error: %s", err)
+			}
+			nodeStatSum := "https://" + ip + ":" + strconv.FormatInt(int64(port), 10) + "/stats/summary"
+			_, err = config.NodeClient.GetRawEndPoint(prefix+n.Name, workDir, nodeStatSum)
+			if err != nil {
+				failedNodeList[n.Name] = err
+				continue
+			}
+			continue
+		}
 
-		nodeStatSum := config.ClusterHostURL + "/api/v1/nodes/" + n + "/proxy/stats/summary"
-		// download node summary
-		_, err := rawClient.GetRawEndPoint(prefix+n, workDir, nodeStatSum)
+		// retrieve node summary via kube-proxy
+		nodeStatSum := config.ClusterHostURL + "/api/v1/nodes/" + n.Name + "/proxy/stats/summary"
+		_, err = config.InClusterClient.GetRawEndPoint(prefix+n.Name, workDir, nodeStatSum)
 		if err != nil {
-			return fmt.Errorf("unable to retrive node: %s %s summary metrics: %s", n, prefix, err)
+			failedNodeList[n.Name] = err
+			continue
 		}
+		continue
+	}
+
+	return failedNodeList, nil
+}
+
+//ensureNodeSource validates connectivity to the cadvisor summary endpoint.
+// If unable to directly connect to the node summary endpoint, attempts to connect via kube-proxy
+func ensureNodeSource(config KubeAgentConfig) (KubeAgentConfig, error) {
+
+	nodeHTTPClient := http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{
+		// nolint gas
+		InsecureSkipVerify: true,
```

Comment:
🍰 👍 unfortunately we have observed that many clusters do not have resolvable DNS entries for their nodes. Because of this we use the IP address to connect to the instance. In the future I would like to make this a user configurable option.